Mathlib/Data/Bitvec/Defs.lean

```diff
@@ -1,339 +1,119 @@
-def Slt (x y : Bitvec n) : Prop :=
-  sborrow x y
-#align bitvec.slt Bitvec.Slt
-
-/-- signed greater-than proposition -/
-def Sgt (x y : Bitvec n) : Prop :=
-  Slt y x
-#align bitvec.sgt Bitvec.Sgt
-
-/-- signed less-than-or-equal-to proposition -/
-def Sle (x y : Bitvec n) : Prop :=
-  ¬Slt y x
-#align bitvec.sle Bitvec.Sle
-
-/-- signed greater-than-or-equal-to proposition -/
-def Sge (x y : Bitvec n) : Prop :=
-  Sle y x
-#align bitvec.sge Bitvec.Sge
-
-end Comparison
-
-/-! ### Conversion to `nat` and `int` -/
-
-
-section Conversion
-
-variable {α : Type}
-
-/-- Create a bitvector from a `nat` -/
-protected def ofNat : ∀ n : ℕ, Nat → Bitvec n
-  | 0, _ => nil
-  | succ n, x => Bitvec.ofNat n (x / 2)++ₜdecide (x % 2 = 1) ::ᵥ nil
-#align bitvec.of_nat Bitvec.ofNat
+/-!
+## Bitwise operations
+-/
 
-/-- Create a bitvector from an `Int`. The ring homomorphism from Int to bitvectors. -/
-protected def ofInt : ∀ n : ℕ, Int → Bitvec n
-  | n, Int.ofNat m => Bitvec.ofNat n m
-  | n, Int.negSucc m => (Bitvec.ofNat n m).not
+/-- Signed less than for bitvectors. -/
+protected def slt (x y : BitVec (w + 1)) : Prop :=
+  ((x.val >>> w = 1) ∧ (y.val >>> w = 0)) ∨ ((x.val >>> w = y.val >>> w) ∧ x < y)
 
-/-- `add_lsb r b` is `r + r + 1` if `b` is `true` and `r + r` otherwise. -/
-def addLsb (r : ℕ) (b : Bool) :=
-  r + r + cond b 1 0
-#align bitvec.add_lsb Bitvec.addLsb
+/-- Signed less than or equal to for bitvectors. -/
+protected def sle (x y : BitVec (w + 1)) : Prop :=
+  ((x.val >>> w = 1) ∧ (y.val >>> w = 0)) ∨ ((x.val >>> w = y.val >>> w) ∧ x ≤ y)
 
-/-- Given a `List` of `Bool`s, return the `nat` they represent as a list of binary digits. -/
-def bitsToNat (v : List Bool) : Nat :=
-  v.foldl addLsb 0
-#align bitvec.bits_to_nat Bitvec.bitsToNat
+/-- Signed greater than for bitvectors. -/
+protected def sgt (x y : BitVec (w + 1)) : Prop := BitVec.slt y x
 
-/-- Return the natural number encoded by the input bitvector -/
-protected def toNat {n : Nat} (v : Bitvec n) : Nat :=
-  bitsToNat (toList v)
-#align bitvec.to_nat Bitvec.toNat
+/-- Signed greater than or equal to for bitvectors. -/
+protected def sge (x y : BitVec (w + 1)) : Prop := BitVec.sle y x
 
-instance (n : ℕ) : Preorder (Bitvec n) :=
-  Preorder.lift Bitvec.toNat
+/-- The left-shift of a bitvector by some amount. -/
+protected def shiftLeft (x : BitVec w) (n : Nat) : BitVec w := (x.val <<< n)
 
-/-- convert `fin` to `Bitvec` -/
-def ofFin {n : ℕ} (i : Fin <| 2 ^ n) : Bitvec n :=
-  Bitvec.ofNat _ i.val
-#align bitvec.of_fin Bitvec.ofFin
+/-- The right-shift of a bitvector by some amount. -/
+protected def shiftRight (x : BitVec w) (n : Nat) : BitVec w :=
+  ⟨x.val >>> n, by
+      simp only [Nat.shiftRight_eq_div_pow]
+      exact lt_of_le_of_lt (Nat.div_le_self' _ _) (x.isLt)⟩
 
-/-- convert `Bitvec` to `fin` -/
-def toFin {n : ℕ} (i : Bitvec n) : Fin (2 ^ n) :=
-  i.toNat
-#align bitvec.to_fin Bitvec.toFin
+instance : Complement (BitVec w) := ⟨(· + (1 : BitVec w))⟩
```

Comment:
This looks wrong to me; should't it be `-1 - x`, so that `0b110 = 6` is sent to `0b001 = 1 = -7`?

Reply:
(Or better, `finReverse` or whatever the existing map is called)